drivers/adodb-postgres64.inc.php

```diff
@@ -272,7 +272,9 @@ function qStr($s, $magic_quotes=false)
 		if ($this->_connectionID) {
 			return "'" . pg_escape_string($this->_connectionID, $s) . "'";
 		} else {
-			return "'" . pg_escape_string($s) . "'";
+			//If there is no database connection, then pg_escape_string() will not work in PHP 8.3+.
+			//  This can occur when using setSessionVariables() in the load balancer, so fall back to emulated escaping in these cases.
+			return parent::qStr( $s );
```

Comment:
I'm OK with this, but can you confirm where you got the information about it no longer working in 8.3 ?

The [pg_escape_string() documentation](https://www.php.net/manual/en/function.pg-escape-string.php) states that _As of PHP 8.1.0, using the default connection is deprecated_. This comes from the [PHP 8.1 deprecations RFC](https://wiki.php.net/rfc/deprecations_php_8_1#calling_overloaded_pgsql_functions_without_the_connection_argument).

Deprecated features are usually removed in the next major release, in that case that would be PHP 9.0.

Reply:
I would need to retest this. I seem to recall it triggering a fatal error, but it may have just been a deprecation warning. I can fix the comment, but either way something is needed here.

Reply:
Actually, I see, the old code was sending the $data argument in the first $connection argument position. PHP 8.2 changed to requiring a PgSQL\Connection instance for that argument, and thus it would fail. I have no idea how it worked at all before to be honest, but it was clearly incorrect.

Reply:
I'm not really sure what *old code* and *$data argument* you are referring to.

Before PHP 8.1, it was possible not to specify the connection ID as first parameter to pg_escape_string(), in which case the default connection (i.e. last used one) was used; this is what was deprecated.

The function's signature also [changed the 1st parameter's type from *resource* to *PgSQL\Connection* object](https://github.com/php/doc-en/commit/c2eca73ef79ebe78cebb34053e41b565af504c4f#diff-3193d27b92fb4a075689186b96ff25aa97d2163286d5c45e24cc387bfbe01a72L16-R16) in 8.1, but the parameter is still marked as optional so other than the deprecation warning there should not be any errors.

Anyway, as mentioned before I'm fine with this change. I've opened #956 to track this.

Reply:
Sorry, I was referring to the argument names that pg_escape_string() accept in the PHP documentation.

 ADODB was only passing a string to escape as the first argument in this case, instead of the expected connection resource or connection object. How it ever functioned is beyond me. But either way it's fixed now.